### DISCUSSION TITLE
Alter positional reference

### DISCUSSION THREAD

**Discussion by trurl-master** [floating-ui/floating-ui#1759] [Q&A] [answered]
Hi! 👋

Thanks for a great lib!

Is there a way to alter the reference but only for positioning?

I have a case (a toolbar) when the floating element appears above a toolbar item on desktop:

<img src="https://user-images.githubusercontent.com/640657/176143336-4ae2d0b9-16c9-455c-9e98-ada6185c3810.png" 
width="350" />

But positioned relative to the toolbar itself on mobile:

<img src="https://user-images.githubusercontent.com/640657/176143487-03bd70ea-3e4b-4721-bcad-7d03bad3202a.png" 
width="250" />

What I basically do now is I have two `useFloating` — one with reference on the item and one on the toolbar, and switch between the two. Something like:

```jsx
const {
  x: posX,
  y: posY,
  reference: posReference,
  floating: posFloating,
} = usePositionalFloating(useFloatingConfig);

const { x, y, reference, floating, strategy } = useFloating(useFloatingConfig);

const floatingRef = useMemo(
  () => mergeRefs([floating, posFloating]),
  [floating, posFloating]
);

useIsomorphicLayoutEffect(() => {
  if (toolbarRef?.current && isPhone) {
    posReference(toolbarRef.current);
  }
}, [isPhone, posReference, toolbarRef]);

return (
  <ul
    ref={floatingRef}
    style={
      isPhone
        ? {
            position: strategy,
            top: posY ?? 0,
            left: posX ?? 0,
          }
         : {
            position: strategy,
            top: y ?? 0,
            left: x ?? 0,
          }
    }
  >...</ul>
  <button ref={reference}>More</button>
)
```

But this is ugly. Is there a more natural way of doing this? Can i hook into `update` somehow and alter the reference?

**Comment by atomiks** [comment, marked as answer]
I'm not sure I understand but couldn't you just use one `useFloating` call and instead merge the reference?

```js
const buttonRef = useRef(null);

const { x, y, reference, floating, strategy } = useFloating(useFloatingConfig);

const referenceRef = useMemo(
  () => mergeRefs([reference, buttonRef]),
  [reference, buttonRef]
);

useIsomorphicLayoutEffect(() => {
  reference(isPhone ? toolbarRef?.current ?? null : buttonRef.current);
}, [isPhone, reference, toolbarRef]);

return (
  <ul
    ref={floating}
    style={{
        position: strategy,
        top: y ?? 0,
        left: x ?? 0,
      }}
  >...</ul>
  <button ref={referenceRef}>More</button>
)
```

**Reply by trurl-master** [reply]
If I understand correctly, that will break interactions, like `useDismiss` won't return focus to the toolbar item, but to the toolbar itself and stuff like that. No?

**Reply by trurl-master** [reply]
I want to design interactions as if the reference is the toolbar item, but position the floating element depending on some condition (`isPhone` in this case)

**Reply by atomiks** [reply]
Ah ok, in the latest version, the interaction hooks only "ignore" a virtual element. 

So what you can do is use a virtual element instead for the `toolbarRef` instead of passing it in directly. Then only the item will be considered the DOM reference for the hooks

```js
reference({
  getBoundingClientRect: () => toolbarRef.current.getBoundingClientRect(),
  contextElement: toolbarRef.current,
});
```

**Reply by trurl-master** [reply]
Not sure i follow, is it something like this?

```jsx
const { x, y, reference, floating, strategy } = useFloating(useFloatingConfig);

useIsomorphicLayoutEffect(() => {
  if (isPhone) {
    reference({
      getBoundingClientRect: () => toolbarRef.current.getBoundingClientRect(),
      contextElement: toolbarRef.current,
    });
  }
}, [...]);

return (
  <ul
    ref={floating}
    style={{
      position: strategy,
      top: y ?? 0,
      left: x ?? 0,
    }}
  >...</ul>
  <button ref={reference}>More</button>
)
```

Will that keep interactions linked to the button, but positioning linked to the toolbar? Calling reference in the effect won't break interactions?

**Reply by atomiks** [reply]
If you're on the latest version 0.6.6. It checks if the value passed in is a real element or not and sets `refs.domReference.current` which is what the hooks use.

```js
  reference(isPhone ? {
      getBoundingClientRect: () => toolbarRef.current.getBoundingClientRect(),
      contextElement: toolbarRef.current,
    } : buttonRef.current);
```

**Reply by trurl-master** [reply]
Yeah, but if I'm on the phone how it will know that the reference for interactions is the button? I feel I'm missing something. Sorry, i just started migrating from Tippy

**Reply by atomiks** [reply]
If you pass `ref={referenceRef}` to the button JSX, it knows to keep using that, despite the effect

**Reply by atomiks** [reply]
Btw, I do need to write docs about this as it's definitely missing

**Reply by trurl-master** [reply]
Yeah, would be great. Thank you!
